### PR TITLE
Plugins Browser: include WP.com features as Jetpack pseudo-plugins

### PR DESCRIPTION
**Pull request by jsnajdr**
When searching for something in Plugins Browser, include matching WP.com features as pseudo-plugins in the search results. The intent is to remind the user that WordPress.com already includes many features that would require either Jetpack or another plugin on a self-hosted site.

wpcom-features-as-plugins

**How to test:**
Select a Simple site and go to the Plugins Browser
1. Click on one of the predefined search terms (Engagement, Writing, ...)
2. Verify that all Jetpack features from that group are displayed as pseudo-plugins (compare with the legacy `/plugins/manage` page for Simple sites)
3. Verify that real WP.org search results are displayed after the Jetpack features and that the infinite scrolling still works (loads more results after scrolling far enough)
4. Enter some other search term, e.g., "SEO"
5. Verify that Jetpack features that have "SEO" in their name or description are displayed in the search results
6. Try to click on any of the pseudo-plugin. You should be navigated to the support page of that feature.

Select a Jetpack/Atomic site, or select "All Sites"
1. Verify that no Jetpack pseudo-plugins are not displayed in this case -- the feature should be limited to Simple sites only.

**How to review:**
There are several preliminary commits that Prettify, ES6-ify and lodash-ify the code. Only the last commit contains the "important" stuff, i.e., the feature itself.

**Open questions (mainly for @iamtakashi)**
The legacy `/plugins/manage` page shows either "Active" for active features, or "Upgrade" button for features that are only in plans I don't have (Premium, Business). Should we display the "Upgrade" button also on the search results page? Right now, I'm showing "Installed" for everything, which should be fixed.

The feature names like "Something Really Nice by Jetpack" don't fit very well into the 3-column view. See the screenshot above.

Should we show Jetpack's rating on every pseudo-plugin? It has "only" 4.5 stars :wink:

The feature's "description" is not displayed anywhere -- we only search in it.

### CODE REVIEW

**Comment by iamtakashi** [comment]
Nice work!!

> Should we display the "Upgrade" button also on the search results page?
Yes!

> The feature names like "Something Really Nice by Jetpack" don't fit very well into the 3-column view. See the screenshot above.
I think we can try 2 column layout between the viewport range of 961 - 1279px.

> Should we show Jetpack's rating on every pseudo-plugin? It has "only" 4.5 stars 
I think it's ok to show it in each :)

> Try to click on any of the pseudo-plugin. You should be navigated to the support page of that feature.
Can you change it to the Jetpack plugin detail page for now? It will probably be better if the outbound links are on the detail page. (We should discuss this another time.)

File: client/my-sites/plugins/plugins-browser-item/index.jsx [hunk 1]
```diff
@@ -102,14 +117,15 @@ const PluginsBrowserListElement = React.createClass( {
 			</li>
 		);
 	}
-} );
+}
 
-export default connect(
-	( state ) => {
+export default compose(
```
Comment (spen) [inline client/my-sites/plugins/plugins-browser-item/index.jsx]: +1 :)

File: client/my-sites/plugins/plugins-browser-item/index.jsx [hunk 2]
```diff
@@ -16,63 +18,68 @@ import Gridicon from 'gridicons';
...
-			iconSize: 40
-		};
-	},
+class PluginsBrowserListElement extends Component {
+	static defaultProps = {
+		iconSize: 40,
+	};
+
+	getPluginLink() {
+		if ( this.props.plugin.link ) {
```
Comment (spen) [inline client/my-sites/plugins/plugins-browser-item/index.jsx]: Any objections to using `get( this.props, 'plugin.link' )` here? We could safely avoid the guarding conditional if so.

Just a drive-by review & suggestion, FFTI :)

**Comment by jsnajdr** [comment]
I added a few changes now:
- if the feature is not available in my plan, we display an "upgrade" button:

wpcom-features-upgrade

I tweaked the button's vertical padding to make it even more compact than a "compact" button. Let me know @iamtakashi if you think it looks nice 😉 

- clicking on the pseudo-plugin now links to the the Jetpack plugin page instead of the support pages. But I'm not very enthusiastic about this change -- I think that showing a support page about a particular feature is much more useful than linking to the generic Jetpack plugin info. Would it be more user-friendly if we opened the support page in a new tab, not leaving Calypso?

There's one more bug I want to fix before merging this -- when infinite scrolling in the list, there's sometimes an infinite loop querying the WP.org API for more and more pages with results.

Also, the changes to the ratings component that @raoulwegat did in #18103 could use some tweaking. The rating stars should be aligned to the bottom of the plugin box (on one line with the button). Also, rating of 82 used to be displayed as 4.5 stars, and now it's just 4. @raoulwegat was that an intentional change? See the screenshots from yesterday (above in the first comment) and today (this comment) to see the difference.

**Comment by jsnajdr** [comment]
> There's one more bug I want to fix before merging this -- when infinite scrolling in the list, there's sometimes an infinite loop querying the WP.org API for more and more pages with results.

That bug was there even before I started making changes in this PR and I'm fixing it separately in #18353.

@iamtakashi @sirbrillig Would you mind giving this one last look before I merge? I'll postpone the suggested 2-column view to another PR. There are test failures reported by CircleCI, but they are harmless -- it's a pre-existing eslint error on a line that I touched by running Prettier.

**Comment by iamtakashi** [comment]
> I tweaked the button's vertical padding to make it even more compact than a "compact" button.

Although it looks good, I don't see it's necessary to introduce a custom style for those button. I'd just use standard compact button.

Everything else is great in design wise 👍

**Comment by jsnajdr** [comment]
I removed the custom styling for the "Upgrade" compact button:

upgrade-plugin

Going to :ship: now!

**Comment by raoulwegat** [comment]
> Also, rating of 82 used to be displayed as 4.5 stars, and now it's just 4. @raoulwegat was that an intentional change?

@jsnajdr I wasn't aware of that. Was that intentional? I was just rounding to what I thought was correct – ie anything above 85 should be 4.5 stars.
